### PR TITLE
refactor: move column name validation out of the parser (MINOR)

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Previously, the parser would throw if C* statements had columns with reserved words or if the key field was not `ROWKEY STRING`.

These checks should not be in the parser, but in the engine as they are requirements of the engine.

This also allows the SchemaParser to handle parsing schemas with multiple key fields, which is needed for the static query work.

### Testing done 

`mvn test`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/TableElements.java [hunk 1]
```diff
@@ -114,17 +113,6 @@ private static TableElements build(final Stream<TableElement> elements) {
     throwOnDuplicateNames(keyColumns, "KEY");
     throwOnDuplicateNames(valueColumns, "non-KEY");
 
-    final long numKeyColumns = keyColumns.size();
```
Comment (vpapavas) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/tree/TableElements.java]: These checks here are not needed?

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CommandFactories.java [hunk 2]
```diff
@@ -265,6 +268,29 @@ private static LogicalSchema buildSchema(final TableElements tableElements) {
       throw new KsqlException("The statement does not define any columns.");
     }
 
+    tableElements.forEach(e -> {
+      if (e.getName().equalsIgnoreCase(SchemaUtil.ROWTIME_NAME)) {
+        throw new KsqlException("'" + e.getName() + "' is a reserved field name.");
+      }
+
+      final boolean isRowKey = e.getName().toUpperCase().equals(SchemaUtil.ROWKEY_NAME);
```
Comment (purplefox) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CommandFactories.java]: For consistency why not use equalsIgnoreCase like on line 272?

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CommandFactories.java [hunk 3]
```diff
@@ -265,6 +268,29 @@ private static LogicalSchema buildSchema(final TableElements tableElements) {
...
+
+      if (e.getNamespace() == Namespace.KEY) {
+        if (!isRowKey) {
+          throw new KsqlException("'" + e.getName() + "' is an invalid KEY field name. "
```
Comment (purplefox) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CommandFactories.java]: Both "field" and "column" seem to be used in this method to represent the same thing (e.g. line 268 uses column, the new code uses field). For consistency I'd suggest choosing one or the other and sticking to that.

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java [hunk 4]
```diff
@@ -825,6 +830,115 @@ public void shouldFailDropSourceOnDropIncompatibleSourceForStream() {
...
+        true,
+        withProperties
+    );
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage("'ROWTIME' is a reserved field name.");
+
+    // When:
+    commandFactories.create("sqlExpresson", statement, ksqlConfig, emptyMap());
```
Comment (purplefox) [inline ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java]: typo? sqlExpression not sqlExpresson

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java [hunk 5]
```diff
@@ -825,6 +830,115 @@ public void shouldFailDropSourceOnDropIncompatibleSourceForStream() {
...
```
Comment (purplefox) [inline ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java]: typo as above

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java [hunk 6]
```diff
@@ -825,6 +830,115 @@ public void shouldFailDropSourceOnDropIncompatibleSourceForStream() {
...
+        SOME_NAME,
+        TableElements.of(tableElement(KEY, ROWKEY_NAME, new Type(SqlTypes.STRING))),
+        true,
+        withProperties
+    );
+
+    // When:
+    commandFactories.create("sqlExpresson", statement, ksqlConfig, emptyMap());
+
+    // Then: did not throw
```
Comment (purplefox) [inline ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java]: If there is a bug in commandFactories.create so it does absolutely nothing this test would still pass. Can we verify somehow that it did what we want it to do?

**Comment by blueedgenick** [comment]
@big-andy-coates just a random question for my own education: why is reserved word handling not a concern of the parser ? (not discussing the ROWKEY part at all, that i 100% agree with). I would have thought that the point of identifying the use of reserved words is that the parser can "confuse itself" with them, which is why it should be the one owning this particular concern ? Suspect i am missing some obvious point here.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/TableElements.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/tree/TableElements.java]: That's why I deleted them ;) :p

So the PR moved these checks, and the ones in `TableElement` out of the parser and into the command factory.   When you combine the checks it's actually the case that these checks are superfluous - the checks in `TableElement` mean its not possible for the data to be in the state these checks detect.

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CommandFactories.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CommandFactories.java]: Yep, will do. (This code was a cut & paste job, but it's a good call)

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CommandFactories.java (hunk 3 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CommandFactories.java]: Yeah, the two are often used interchangeably. However, I think it's cleaner to use `column` when referring to a column in the table and `field` only for fields within a `STRUCT`.

I'll update all to `column` in this instance.

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java (hunk 4 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java]: done

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java (hunk 5 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java]: done

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java (hunk 6 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java]: You've picked up on this in other PRs as well.  I disagree with your reasoning.  The point of this test is to prove it does not throw an exception under a certain circumstance.  Yes, an empty implementation, or one with a bug, might pass this test.  But there should be other tests that cover that and they should fail.  This test is only testing that no exception is thrown. It is a valid test.

**Comment by big-andy-coates** [comment]
> @big-andy-coates just a random question for my own education: why is reserved word handling not a concern of the parser ? (not discussing the ROWKEY part at all, that i 100% agree with). I would have thought that the point of identifying the use of reserved words is that the parser can "confuse itself" with them, which is why it should be the one owning this particular concern ? Suspect i am missing some obvious point here.

The parse will handle its own concept of reserved words, e.g. you can't have a column named `FROM` unless you quote it.    This is what you're alluding to, and it hasn't changed or moved.

What this PR does is move some additional, secondary, validation of column names out of the parser.  This validation is required because the _engine_ can't handle CT/CS statements that have a column called `ROWTIME` or a column called `ROWKEY`.   As these are _engine_ requirements it makes sense to move the validation to the engine.

**Pull request merged**